email/model/jpa/src/main/java/com/blazebit/notify/email/model/jpa/AbstractEmailNotification.java

```diff
@@ -39,66 +36,73 @@
 import javax.persistence.MappedSuperclass;
 import javax.persistence.Transient;
 import javax.validation.constraints.NotNull;
-import java.time.Instant;
-import java.util.Locale;
-import java.util.Map;
 
 /**
  * An abstract mapped superclass implementing the {@link Notification} interface for E-Mail notifications.
  *
  * @param <ID> The id type of the notification
+ * @param <F> The type representing the email sender
  * @author Christian Beikov
  * @since 1.0.0
  */
 @MappedSuperclass
-public abstract class AbstractEmailNotification<ID> extends AbstractNotification<ID> implements ConfigurationSourceProvider {
+public abstract class AbstractEmailNotification<ID, F extends AbstractFromEmail> extends AbstractNotification<ID> implements ConfigurationSourceProvider {
 
-    /**
-     * The parameter name for the template processor type {@link TemplateProcessorKey#getTemplateProcessorResultType()}.
-     */
-    public static final String TEMPLATE_PROCESSOR_TYPE_PARAMETER_NAME = "templateProcessorType";
     /**
      * The parameter name for the subject which is a plain {@link String}.
      * The value represents the literal subject.
      */
-    public static final String SUBJECT_PARAMETER_NAME = "subject";
+    public static final String SUBJECT_PARAMETER = "subject";
     /**
      * The parameter name for the subject template which is a plain {@link String}.
      * The value refers to a template name.
      */
-    public static final String SUBJECT_TEMPLATE_PARAMETER_NAME = "subjectTemplate";
+    public static final String SUBJECT_TEMPLATE_PARAMETER = "subjectTemplate";
+    /**
+     * The parameter name for the subject template processor key which is a plain {@link String}.
+     * The value refers to a template processor key.
+     */
+    public static final String SUBJECT_TEMPLATE_PROCESSOR_KEY_PARAMETER = "subjectTemplateProcessorKey";
```

Comment:
I think this is a horrible pattern. People will usually only use a single template engine. So storing the template engine name for every email is just useless. The template engine should be resolved based on some resolver the user can override, which may receive access to the notification itself, so that the user can put the template engine name somehwere in case it isn't always the same.

We have to rethink this part a bit. We need something like this:

```java
interface TemplateProcessorFactoryResolver {
    TemplateProcessorFactory<T> resolveTemplateProcessorFactory(TemplateContext context, Class<T> resultType, Notification<?> notification, ConfigurationSource confiurationSource, ServiceProvider serviceProvider);
}
```

Reply:
I think this interface is not needed as the user could just create a custom TemplateProcessorFactory that performs some kind of delegation. So maybe let's reduce the discussion to only allowing a single TemplateProcessorFactory per notification.

The issue is a different one: It must be possible to use different template processors for different parts of _the same_ notification. Not sure how this interface would make this possible? If you only allow a single TemplateProcessorFactory then the caller must somehow provide information for what part of the notification the template processor is being created. I don't know how to do this in a generic fashion.

Reply:
Well, we can do it that way but it basically shifts the responsibility to the user having to implement custom `TemplateProcessorFactory` where delegation is performed based on the `templateName`.

Reply:
IMO the user of blaze-notify should at most configure a base path where to find templates on the class path, but even for that I would suggest we use some sane default like e.g. `templates/`